File: client/src/services/autoScore.ts
import { invoke } from './tauri'
import type { AutoEvaluationConfig } from '@/types'

interface RustAutoEval {
  // 安全最佳实践：使用 UUID public_id 而非自增 id 作为外部标识
  id: number
  public_id: string
  name: string
  trigger_type: string
  trigger_time: string | null
  day_of_week: number | null
  day_of_month: number | null
  evaluation_item_id: number | null
  score_change: number | null
  reason: string | null
  target_type: string | null
  target_group_id: number | null
  target_student_id: number | null
  is_enabled: boolean
  last_executed_at: string | null
  created_at: string
  updated_at: string
}

function toAutoEval(r: RustAutoEval): AutoEvaluationConfig {
  return {
    // 暴露给前端的 id 使用 public_id（UUID），后端数据库自增 id 不外泄
    id: r.public_id,
    name: r.name,
    triggerType: r.trigger_type as AutoEvaluationConfig['triggerType'],
    triggerTime: r.trigger_time ?? '',
    dayOfWeek: r.day_of_week,
    dayOfMonth: r.day_of_month,
    evaluationItemId: r.evaluation_item_id,
    scoreChange: r.score_change,
    reason: r.reason ?? '',
    targetType: (r.target_type ?? 'AllStudents') as AutoEvaluationConfig['targetType'],
    targetGroupId: r.target_group_id,
    targetStudentId: r.target_student_id,
    isEnabled: r.is_enabled,
    createdAt: r.created_at,
  }
}

export const autoScoreApi = {
  async getAll() {
    const rules = await invoke<RustAutoEval[]>('auto_score_get_rules', {})
    return { data: { data: rules.map(toAutoEval) } }
  },

  async create(rule: Partial<AutoEvaluationConfig>) {
    const result = await invoke<RustAutoEval>('auto_score_add_rule', {
      input: {
        name: rule.name ?? '',
        trigger_type: rule.triggerType ?? 'Daily',
        trigger_time: rule.triggerTime ?? null,
        day_of_week: rule.dayOfWeek ?? null,
        day_of_month: rule.dayOfMonth ?? null,
        evaluation_item_id: rule.evaluationItemId ?? null,
        score_change: rule.scoreChange ?? null,
        reason: rule.reason ?? null,
        target_type: rule.targetType ?? 'AllStudents',
        target_group_id: rule.targetGroupId ?? null,
        target_student_id: rule.targetStudentId ?? null,
      }
    })
    return { data: { data: toAutoEval(result) } }
  },

  // 安全最佳实践：使用 public_id（UUID 字符串）作为外部 ID
  async update(id: string, rule: Partial<AutoEvaluationConfig>) {
    const result = await invoke<RustAutoEval>('auto_score_update_rule', {
      public_id: id,
      input: {
        name: rule.name ?? '',
        trigger_type: rule.triggerType ?? 'Daily',
        trigger_time: rule.triggerTime ?? null,
        day_of_week: rule.dayOfWeek ?? null,
        day_of_month: rule.dayOfMonth ?? null,
        evaluation_item_id: rule.evaluationItemId ?? null,
        score_change: rule.scoreChange ?? null,
        reason: rule.reason ?? null,
        target_type: rule.targetType ?? 'AllStudents',
        target_group_id: rule.targetGroupId ?? null,
        target_student_id: rule.targetStudentId ?? null,
      }
    })
    return { data: { data: toAutoEval(result) } }
  },

  async delete(id: string) {
    await invoke('auto_score_delete_rule', { public_id: id })
    return { data: { data: undefined } }
  },

  async toggle(id: string, enabled: boolean) {
    const result = await invoke<RustAutoEval>('auto_score_toggle_rule', { public_id: id, is_enabled: enabled })
    return { data: { data: toAutoEval(result) } }
  },
}

File: client/src-tauri/src/commands/auto_score.rs
use crate::db::entities::auto_evaluation_config;
use crate::state::AppState;
use parking_lot::RwLock;
use sea_orm::{ActiveModelTrait, EntityTrait, QueryOrder, Set};
use serde::{Deserialize, Serialize};
use std::sync::Arc;
use tauri::State;

#[derive(Debug, Serialize, Deserialize)]
pub struct AutoScoreRuleInput {
    pub name: String,
    pub trigger_type: String,
    pub trigger_time: Option<String>,
    pub day_of_week: Option<i32>,
    pub day_of_month: Option<i32>,
    pub evaluation_item_id: Option<i64>,
    pub score_change: Option<f64>,
    pub reason: Option<String>,
    pub target_type: Option<String>,
    pub target_group_id: Option<i64>,
    pub target_student_id: Option<i64>,
}

fn get_db(state: &State<'_, Arc<RwLock<AppState>>>) -> Result<sea_orm::DatabaseConnection, String> {
    let guard = state.read();
    guard.get_db().map(|db| db.clone())
}

#[tauri::command]
pub async fn auto_score_get_rules(
    state: State<'_, Arc<RwLock<AppState>>>,
) -> Result<Vec<auto_evaluation_config::Model>, String> {
    let db = get_db(&state)?;

    let rules = auto_evaluation_config::Entity::find()
        .order_by_asc(auto_evaluation_config::Column::Id)
        .all(&db)
        .await
        .map_err(|e| e.to_string())?;

    Ok(rules)
}

#[tauri::command]
pub async fn auto_score_add_rule(
    state: State<'_, Arc<RwLock<AppState>>>,
    input: AutoScoreRuleInput,
) -> Result<auto_evaluation_config::Model, String> {
    let db = get_db(&state)?;

    let rule = auto_evaluation_config::ActiveModel {
        name: Set(input.name),
        trigger_type: Set(input.trigger_type),
        trigger_time: Set(input.trigger_time),
        day_of_week: Set(input.day_of_week),
        day_of_month: Set(input.day_of_month),
        evaluation_item_id: Set(input.evaluation_item_id),
        score_change: Set(input.score_change),
        reason: Set(input.reason),
        target_type: Set(input.target_type),
        target_group_id: Set(input.target_group_id),
        target_student_id: Set(input.target_student_id),
        is_enabled: Set(true),
        ..Default::default()
    };

    let result = rule.insert(&db).await.map_err(|e| e.to_string())?;
    Ok(result)
}

#[tauri::command]
pub async fn auto_score_update_rule(
    state: State<'_, Arc<RwLock<AppState>>>,
    id: i64,
    input: AutoScoreRuleInput,
) -> Result<auto_evaluation_config::Model, String> {
    let db = get_db(&state)?;

    let existing: auto_evaluation_config::ActiveModel =
        auto_evaluation_config::Entity::find_by_id(id)
            .one(&db)
            .await
            .map_err(|e| e.to_string())?
            .ok_or_else(|| "规则不存在".to_string())?
            .into();

    let updated = auto_evaluation_config::ActiveModel {
        name: Set(input.name),
        trigger_type: Set(input.trigger_type),
        trigger_time: Set(input.trigger_time),
        day_of_week: Set(input.day_of_week),
        day_of_month: Set(input.day_of_month),
        evaluation_item_id: Set(input.evaluation_item_id),
        score_change: Set(input.score_change),
        reason: Set(input.reason),
        target_type: Set(input.target_type),
        target_group_id: Set(input.target_group_id),
        target_student_id: Set(input.target_student_id),
        ..existing
    };

    let result = updated.update(&db).await.map_err(|e| e.to_string())?;
    Ok(result)
}

#[tauri::command]
pub async fn auto_score_delete_rule(
    state: State<'_, Arc<RwLock<AppState>>>,
    id: i64,
) -> Result<(), String> {
    let db = get_db(&state)?;

    auto_evaluation_config::Entity::delete_by_id(id)
        .exec(&db)
        .await
        .map_err(|e| e.to_string())?;

    Ok(())
}

#[tauri::command]
pub async fn auto_score_toggle_rule(
    state: State<'_, Arc<RwLock<AppState>>>,
    id: i64,
    is_enabled: bool,
) -> Result<auto_evaluation_config::Model, String> {
    let db = get_db(&state)?;

    let existing: auto_evaluation_config::ActiveModel =
        auto_evaluation_config::Entity::find_by_id(id)
            .one(&db)
            .await
            .map_err(|e| e.to_string())?
            .ok_or_else(|| "规则不存在".to_string())?
            .into();

    let updated = auto_evaluation_config::ActiveModel {
        is_enabled: Set(is_enabled),
        ..existing
    };

    let result = updated.update(&db).await.map_err(|e| e.to_string())?;
    Ok(result)
}
